lib/internal/streams/end-of-stream.js

```diff
@@ -71,6 +80,18 @@ function eos(stream, opts, callback) {
     stream.req.on('finish', onfinish);
   };
 
+  const isEmittingClose = isStreamEmittingClose(stream);
+
+  if (!isEmittingClose) {
+    const _destroy = stream._destroy;
+    stream._destroy = function(err, cb) {
+      _destroy.call(stream, err, (_err) => {
+        if (!_err) process.nextTick(() => callback.call(stream, _err));
+        cb(_err);
+      });
+    };
+  }
```

Comment:
I do not think this is the correct fix. We should not monkey patching the incoming stream. If the default implementation of destroy needs a change, let's do it there.

Reply:
I think that the main problem here is that currently it's allowed for stream to be silently destroyed, but we can not change this behaviour now, right?

Reply:
The stream should not be tampered with in any form apart from adding events. This is a hard rule, this function is not expected to alter the internals in the streams. We should also not rely on internals as much as possible, as "previous generation" streams could be passed inside these.

Reply:
Ok, I can also think about introducing new “destroy” event which will be always emitted after “.destroy()” method.

Would it be the better way? My initial idea was to touch the smallest piece of code and functionality as possible. Each stream has “.destroy()” method so it seems natural to monkey patch it. In this way we don’t even need to introduce something new.

Reply:
I don't think this is the right fix for https://github.com/nodejs/node/issues/26550, take a look at https://github.com/nodejs/node/issues/26550#issuecomment-473355056.
Essentially it's process.stdout that has a bug, not pipeline. We should better document the expectations of a `_destroy` method, which are violated by that.

Reply:
But it’s not a bug in a sense: “process.stdout” is a stream that has an option “emitClose:false” in their constructor which leads to this behavior. “end-of-stream” actually has a bug, it can not catch such streams.

Reply:
end-of-stream cannot catch those because `emitClose: false` prevent them to be observable. That's the whole reason for that option - it's a "you should handle this yourself if you want to, these are dangerous waters" option. Now, the documentation might be improved..

Reply:
why does stdout set emitClose to false?

Reply:
`net` set it to false because it handles `emit('close')` in its own `_destroy` method. However in https://github.com/nodejs/node/blob/1cdeb9f956a9f54dde7c6b1a792b97584acbfb8e/lib/internal/process/stdio.js#L17 we override `_destroy`, and that means `'close'` is never emitted.
I have a PR coming.

Reply:
Generically a stream that sets `emitClose: false` is responsible for emitting close in user-logic.

Reply:
yep, that sounds reasonable!